module/VuFind/src/VuFind/Controller/CheckoutsController.php

```diff
@@ -0,0 +1,290 @@
+        $result
+            = $catalog->getMyTransactionHistory($patron, $pageOptions['ilsParams']);
+
+        if (isset($result['success']) && !$result['success']) {
+            $this->flashMessenger()->addErrorMessage($result['status']);
+            return $this->createViewModel();
+        }
+
+        $paginator = $this->paginationHelper->getPaginator(
+            $pageOptions,
+            $result['count'],
+            $result['transactions']
+        );
+        if ($paginator) {
+            $pageStart = $paginator->getAbsoluteItemNumber(1) - 1;
+            $pageEnd = $paginator->getAbsoluteItemNumber($pageOptions['limit']) - 1;
+        } else {
+            $pageStart = 0;
+            $pageEnd = $result['count'];
+        }
+
+        $driversNeeded = $hiddenTransactions = [];
+        foreach ($result['transactions'] as $i => $current) {
+            // Build record drivers (only for the current visible page):
+            if ($pageOptions['ilsPaging'] || ($i >= $pageStart && $i <= $pageEnd)) {
+                $driversNeeded[] = $current;
+            } else {
+                $hiddenTransactions[] = $current;
+            }
+            if ($purgeSelectedAllowed && isset($current['row_id'])) {
+                $this->rememberValidRowId($current['row_id']);
+            }
+        }
+
+        $transactions = $this->ilsRecords()->getDrivers($driversNeeded);
+        $sortList = $pageOptions['sortList'];
+        $params = $pageOptions['ilsParams'];
+        return $this->createViewModel(
+            compact(
+                'transactions',
+                'paginator',
+                'params',
+                'hiddenTransactions',
+                'sortList',
+                'functionConfig',
+                'purgeAllAllowed',
+                'purgeSelectedAllowed'
+            )
+        );
+    }
+
+    /**
+     * Purge loans from loan history
+     *
+     * @return mixed
+     */
+    public function purgeHistoryAction()
+    {
+        $this->ilsExceptionResponse = $redirectResponse
+            = $this->redirect()->toRoute('checkouts-history');
+
+        // Stop now if the user does not have valid catalog credentials available:
+        if (!is_array($patron = $this->catalogLogin())) {
+            return $patron;
+        }
+
+        if (
+            $this->formWasSubmitted('purgeSelected', false)
+            || $this->formWasSubmitted('purgeAll', false)
+        ) {
+            $csrfToken = $this->getRequest()->getPost()->get('csrf');
+            if (!$this->csrf->isValid($csrfToken)) {
+                $this->flashMessenger()
+                    ->addErrorMessage('error_inconsistent_parameters');
+                return $redirectResponse;
+            }
+            // After successful token verification, clear list to shrink session:
+            $this->csrf->trimTokenList(0);
+            $catalog = $this->getILS();
+            if ($this->formWasSubmitted('purgeAll', false)) {
```

Comment:
Would it be worth saving this result in a variable so we don't make the same formWasSubmitted call twice?

Reply:
Good idea. I put them both in variables for uniformity.